Part 2: Add signal sensor

ocalvo opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
No
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->

## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This change adds:
  - Incoming SMS notifications.
  - Signal sensor to track the GSM signal, usefull to diagnose failures when sending SMS messages.
  - Config flow support

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->
N/A

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Also implements config flow support

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/[updated](https://github.com/home-assistant/home-assistant.io/pull/13067) for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

ocalvo committed:
Add GSM signal sensor

frenck commented:
We don't accept or allow for integration to be configured on platform level.

See ADR-0007: <https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md>

Also, the recently introduced ADR-0010 probably applies here as well

<https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md>

ocalvo committed:
Add config flow support

ocalvo commented:
> We don't accept or allow for integration to be configured on platform level.
> 
> See ADR-0007: https://github.com/home-assistant/architecture/blob/master/adr/0007-integration-config-yaml-structure.md
> 
> Also, the recently introduced ADR-0010 probably applies here as well
> 
> https://github.com/home-assistant/architecture/blob/master/adr/0010-integration-configuration.md

Added support for config flow and removed the need for users to edit yaml.

ocalvo committed:
Add missing files; minor fix

ocalvo commented:
Hi @frenck , do you have time to review this one?

frenck commented:
@ocalvo Please don't ping people to get a review. Thanks 👍

ocalvo committed:
Merge branch 'dev' of https://github.com/home-assistant/core into AddGSMSignalSensor

ocalvo committed:
Use new translation folder

ocalvo committed:
Merge branch 'dev' of https://github.com/home-assistant/core into AddGSMSignalSensor

ocalvo committed:
Merge branch 'dev' of https://github.com/home-assistant/core into AddGSMSignalSensor

ocalvo committed:
Merge branch 'dev' of https://github.com/home-assistant/core into AddGSMSignalSensor

ocalvo committed:
Add async worker

ocalvo committed:
Fix terminate bug

ocalvo committed:
Fire events for incomming SMS

MartinHjelmare reviewed (changes requested):
Please split this PR into at least 3 separate PRs. One per bullet:

- Config flow
- Incoming SMS notifications
- Signal sensor

ocalvo committed:
Make pylint clean

ocalvo commented:
> Please split this PR into at least 3 separate PRs. One per bullet:
> 
> * Config flow
> * Incoming SMS notifications
> * Signal sensor

Sure thing. I will first make the entire PR pylint clean and re-test the end to end experience to make sure everything is working.
Then I will send 3 separate PRs.

ocalvo committed:
Make pylint clean

ocalvo committed:
Refactor prep

ocalvo commented:
> > Please split this PR into at least 3 separate PRs. One per bullet:
> > 
> > * Config flow
> > * Incoming SMS notifications
> > * Signal sensor
> 
> Sure thing. I will first make the entire PR pylint clean and re-test the end to end experience to make sure everything is working.
> Then I will send 3 separate PRs.

[Config flow](https://github.com/home-assistant/core/pull/35995)

ocalvo committed:
Merge branch 'dev' of https://github.com/home-assistant/core into AddGSMSignalSensor

ocalvo commented:
Incoming SMS is currently [blocked in hassio](https://github.com/home-assistant/docker/issues/119) by this [gammu issue](https://github.com/gammu/gammu/issues/460)

ocalvo committed:
Merge branch 'dev' into AddGSMSignalSensor

ocalvo committed:
Merge branch 'AddGSMSignalSensor' of https://github.com/ocalvo/home-assistant into AddGSMSignalSensor

ocalvo committed:
Remove not needed file

ocalvo committed:
Add translations

ocalvo committed:
Touch files

ocalvo committed:
Merge branch 'dev' into AddGSMSignalSensor

balloob commented on homeassistant/components/sms/sensor.py:
```diff
@@ -0,0 +1,78 @@
+    async_add_entities(entities, True)
+
+
+class GSMSignalSensor(Entity):
```
These kind of sensors tend to be very spammy, and generally don't have too much added value. I suggest you disable it by default by implementing the `entity_registry_enabled_default` property and defaulting it to `False`

ocalvo replied:
👍

balloob commented on homeassistant/components/sms/sensor.py:
```diff
@@ -0,0 +1,78 @@
+        """Return the state of the device."""
+        return self._state["SignalStrength"]
+
+    @property
```
This is not a property.

balloob commented on homeassistant/components/sms/sensor.py:
```diff
@@ -0,0 +1,78 @@
+        return self._state["SignalStrength"]
+
+    @property
+    async def async_force_update(self):
```
Who calls this function ?

balloob commented on homeassistant/components/sms/sensor.py:
```diff
@@ -0,0 +1,78 @@
+    async_add_entities(entities, True)
+
+
+class GSMSignalSensor(Entity):
```
Please also include a unique ID property. If you don't have access to a unique identifier, use 

```python
f"{config_entry.entry_id}_signal"
```

balloob commented on homeassistant/components/sms/__init__.py:
```diff
@@ -44,13 +47,27 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
     if not gateway:
         return False
     hass.data[DOMAIN][SMS_GATEWAY] = gateway
```
Don't use a constant but instead use `entry.entry_id` to store the gateway. That way you can support multiple config flows.

ocalvo replied:
> Don't use a constant but instead use `entry.entry_id` to store the gateway. That way you can support multiple config flows.

We cannot use the entry id as we need to obtain the gateway from the notify service, but notify services does not support config flow yet.

balloob replied:
Ah, too bad 😢

ocalvo committed:
Apply PR feedback

ocalvo committed:
Merge branch 'AddGSMSignalSensor' of https://github.com/ocalvo/home-assistant into AddGSMSignalSensor

MartinHjelmare commented on homeassistant/components/sms/sensor.py:
```diff
@@ -0,0 +1,76 @@
+    """Implementation of a GSM Signal sensor."""
+
+    def __init__(
+        self, hass, gateway, name,
```
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant.

MartinHjelmare reviewed (commented):
Please address the comment in a new PR.
